File: 프로그래머스/lv2/42890. 후보키/후보키.py
combinations = []

def solution(relation):
    indexList = list(range(0, len(relation[0])))
    visited = [False] * (len(indexList) + 1)
    for i in range(1, len(indexList)+1):
        makeCombination(indexList, visited, 0, len(indexList), i)

    new_comb = combinations
    answer = []
    while (len(new_comb) != 0):
        c = new_comb.pop(0)
        if (canBeKey(relation, c)):
            answer.append(c)
            new_comb = [value for value in new_comb if not all(item in value for item in c)]

    return len(answer)

def makeCombination(arr, visited, start, n, r):
    if (r == 0):
        combinations.append(addOn(arr, visited))
        return

    for i in range(start, len(arr)):
        visited[i] = True
        makeCombination(arr, visited, i + 1, n, r - 1)
        visited[i] = False

def addOn(arr, visited):
    array = []
    for i in range(0, len(arr)):
        if (visited[i]):
            array.append(arr[i])
    return array

def canBeKey(relation, keyIndex):
    s = list()

    for r in relation:
        sub = [r[i] for i in keyIndex]

        if (sub in s):
            return False
        else:
            s.append(sub)
    return True

File: 프로그래머스/lv2/42890. 후보키/후보키.java
import java.util.*;

class Solution {
    static List<List<Integer>> combinations;
    public int solution(String[][] relation){
        combinations = new ArrayList<>();
        List<Integer> indexList = new ArrayList<>();
        for(int i = 0; i < relation[0].length; i++){
            indexList.add(i);
        }

        boolean[] visited = new boolean[indexList.size() + 1];
        for(int i = 1; i <= indexList.size(); i++){
            makeCombination(indexList, visited, 0, indexList.size(), i);
        }

        List<Integer> answer = new ArrayList<>();
        while(combinations.size() != 0){
            List<Integer> c = combinations.remove(0);
            if(canBeKey(relation, c)){
                answer.add(c.size());
                List<List<Integer>> newCombinations = new ArrayList<>();
                for(List<Integer> value : combinations){
                    boolean isSubSet = true;
                    for(int i = 0; i < c.size(); i++){
                        if(!value.contains(c.get(i))){
                            isSubSet = false;
                            break;
                        }
                    }
                    if(!isSubSet){
                        newCombinations.add(value);
                    }
                }
                combinations = newCombinations;
            }
            
        }
        return answer.size();
    }

    static void makeCombination(List<Integer> arr, boolean[] visited, int start, int n, int r){
        if(r == 0){
            combinations.add(addOn(arr, visited));
            return;
        }

        for(int i = start; i < arr.size(); i++){
            visited[i] = true;
            makeCombination(arr, visited, i + 1, n, r - 1);
            visited[i] = false;
        }
    }
    static List<Integer> addOn(List<Integer> arr, boolean[] visited){
        List<Integer> result = new ArrayList<>();
        for(int i = 0; i < arr.size(); i++){
            if(visited[i]){
                result.add(arr.get(i));
            }
        }
        return result;
    }
    static boolean canBeKey(String[][] relation, List<Integer> comb){
        List<String> keyList = new ArrayList<>();
        for(int i = 0; i < relation.length; i++){
            String key = "";
            for(int j = 0; j < comb.size(); j++){
                key += relation[i][comb.get(j)];
            }
            if(keyList.contains(key)){
                return false;
            }
            keyList.add(key);
        }
        return true;
    }
}